### PR TITLE
restructure things and add comments

### PR DESCRIPTION
**Pull request by raulk**
- state load and store methods are now members of the `State` struct.
- dispatch table is cleaner.
- added comments explaining how DX could improve.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -68,34 +88,21 @@ pub fn invoke(_: u32) -> u32 {
...
 }
 
+/// The constructor populates the initial state.
+///
+/// Method num 1. This is part of the Filecoin calling convention.
+/// InitActor#Exec will call the constructor on method_num = 1.
 pub fn constructor() -> Option<RawBytes> {
     let state = State::default();
-    save_state(&state);
+    state.save();
```
Comment (vyzo) [inline src/lib.rs:108]: a TODO about the reset bug would be useful, it is still here.

**Review by vyzo — APPROVED** [review]
much nicer!

We just need a TODO/warning about the multiple constructor invocation bug.

File: src/lib.rs (hunk 1 above)
Comment (raulk) [reply src/lib.rs:108]: Oh yeah, let me add an ACL check here.

Comment (raulk) [reply src/lib.rs:108]: https://github.com/raulk/fil-hello-world-actor/pull/3/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R97-R104

Comment (raulk) [reply src/lib.rs:108]: https://github.com/raulk/fil-hello-world-actor/pull/3/commits/1b8aaab57ca82d296c041752b364c6685d57e84d

Comment (raulk) [reply src/lib.rs:108]: [`1b8aaab` (#3)](https://github.com/raulk/fil-hello-world-actor/pull/3/commits/1b8aaab57ca82d296c041752b364c6685d57e84d)

**Pull request merged**